### ISSUE TITLE
Incorrect handling of SequenceReset Gap Fill Message

### ISSUE THREAD

**Issue by kmdshuaib** [connamara/quickfixn#309]
Hi All,

While performing a message recovery operation in chunks of 2500 messages per resend request,
drop copy gap fills the administrative messages.

Assuming sending "resend request’s” from 10260 till 36264.

1st resend request 10260 till 12759
2nd resend request 12760 till 15259
3rd resend request 15260 till 17759
4th  resend request 17760 till 20259

And so on……

While responding the 3rd resend request, the counter party sends the SequenceReset as there was a gap fill for administrative messages.

However the NewSequence number in that SequenceReset is 36265 (that is actually greater than 3rd resend request end sequence no 17759).

And then QUICK FIX ignores the next expected resend request itself i.e., 4th  resend request 17760 till 20259

How should we handle this scenario as such to send the 4th resend request when there is a gapfill towards the end of the resend request response?

Please advice.

**Comment by dungvu-qe** [comment]
We have same issue, :(

**Comment by dungvu-qe** [comment]
Here is all msg from server

2015/05/27 17:49:07:836: FIXPump: FIX session established on connection: QECAP
2015/05/27 17:49:07:837: FIXFormatter_QECAP: SENDING RESEND REQUEST - seq# begin: 4026 end: 4034 current: 4026 to: QECAP
2015/05/27 17:49:07:837: FIXConnectionData: Sending data on connection {QECAP} [8=FIX.4.4^A9=0074^A35=2^A49=...^A56=...^A34=4123^A52=20150527-09:49:07^A7=4026^A16=4034^A10=152^A]

2015/05/27 17:49:07:913: FIXPump: Received data on connection {QECAP} [8=FIX.4.4^A9=108^A35=4^A34=4026^A43=Y^A49=...^A52=20150527-09:49:03.971^A56=...^A122=20150527-09:49:03.971^A36=4035^A123=Y^A10=052^A]
2015/05/27 17:49:07:913: FIXPump:QECAP: RECEIVED SequenceReset from 4026 to 4035
2015/05/27 17:49:07:913: FIXPump:QECAP: RECV - Resend Request Done received end seq no 4034

**Comment by hungle-qe** [comment]
message log client side:

20150527-09:45:37.052 : 8=FIX.4.49=9135=034=402549=Client52=20150527-09:45:37.05256=Server112=Appia-20150527-09:45:4010=160
20150527-09:45:40.313 : 8=FIX.4.49=23535=D34=402649=Client52=20150527-09:45:40.31356=Server1=CS555A11=19736034005115=USD21=122=538=440=244=126.348=173049=Client54=255=KC59=0107=3115=116=Client200=201507207=IEPA461=FXXXXX513=TCB03800110=163
20150527-09:46:12.043 : 8=FIX.4.49=7535=A34=402749=Client52=20150527-09:46:12.04356=Server98=0108=1010=111
20150527-09:46:42.463 : 8=FIX.4.49=7535=A34=402849=Client52=20150527-09:46:42.46356=Server98=0108=1010=121
20150527-09:46:46.691 : 8=FIX.4.49=23535=D34=402949=Client52=20150527-09:46:46.69156=Server1=CS555A11=19736040644515=USD21=122=538=340=244=163348=177949=Client54=155=LRC59=0107=3115=116=Client200=201511207=IEPA461=FXXXXX513=TCB01100110=227
20150527-09:47:12.883 : 8=FIX.4.49=7535=A34=403049=Client52=20150527-09:47:12.88356=Server98=0108=1010=118
20150527-09:47:43.303 : 8=FIX.4.49=7535=A34=403149=Client52=20150527-09:47:43.30356=Server98=0108=1010=110
20150527-09:48:13.723 : 8=FIX.4.49=7535=A34=403249=Client52=20150527-09:48:13.72356=Server98=0108=1010=115
20150527-09:48:13.801 : 8=FIX.4.49=007135=A49=Server56=Client34=411952=20150527-09:48:1798=0108=1010=015
20150527-09:48:13.801 : 8=FIX.4.49=7535=234=403349=Client52=20150527-09:48:13.80156=Server7=411516=010=096
20150527-09:48:13.801 : 8=FIX.4.49=007435=249=Server56=Client34=412052=20150527-09:48:177=402216=403110=142
20150527-09:48:13.801 : 8=FIX.4.49=10835=434=402243=Y49=Client52=20150527-09:48:13.80156=Server122=20150527-09:48:13.80136=4026123=Y10=032
20150527-09:48:13.801 : 8=FIX.4.49=26635=D34=402643=Y49=Client52=20150527-09:48:13.80156=Server115=116=Client122=20150527-09:45:40.3131=CS555A11=19736034005115=USD21=122=538=440=244=126.348=173054=255=KC59=0107=3200=201507207=IEPA461=FXXXXX513=TCB03800110=175
20150527-09:48:13.801 : 8=FIX.4.49=10835=434=402743=Y49=Client52=20150527-09:48:13.80156=Server122=20150527-09:48:13.80136=4029123=Y10=040
20150527-09:48:13.801 : 8=FIX.4.49=26635=D34=402943=Y49=Client52=20150527-09:48:13.80156=Server115=116=Client122=20150527-09:46:46.6911=CS555A11=19736040644515=USD21=122=538=340=244=163348=177954=155=LRC59=0107=3200=201511207=IEPA461=FXXXXX513=TCB01100110=239
20150527-09:48:13.801 : 8=FIX.4.49=10835=434=403043=Y49=Client52=20150527-09:48:13.80156=Server122=20150527-09:48:13.80136=4032123=Y10=028
20150527-09:48:13.833 : 8=FIX.4.49=010035=449=Server56=Client34=411543=Y122=20150527-09:48:1752=20150527-09:48:1736=4121123=Y10=193
20150527-09:48:13.833 : 8=FIX.4.49=010535=349=Server56=Client34=412152=20150527-09:48:1745=402658=Framing error: Checksum not found.10=117
20150527-09:49:03.909 : 8=FIX.4.49=7535=A34=403549=Client52=20150527-09:49:03.87756=Server98=0108=1010=128
20150527-09:49:03.940 : 8=FIX.4.49=007135=A49=Server56=Client34=412252=20150527-09:49:0798=0108=1010=009
20150527-09:49:03.971 : 8=FIX.4.49=007435=249=Server56=Client34=412352=20150527-09:49:077=402616=403410=152
20150527-09:49:03.971 : 8=FIX.4.49=10835=434=402643=Y49=Client52=20150527-09:49:03.97156=Server122=20150527-09:49:03.97136=4035123=Y10=052
20150527-09:49:04.080 : 8=FIX.4.49=23535=D34=403649=Client52=20150527-09:49:04.08056=Server1=CS555A11=19736053985815=USD21=122=538=140=244=150.548=173049=Client54=255=KC59=0107=3115=116=Client200=201507207=IEPA461=FXXXXX513=TCB03800110=190

**Comment by dungvu-qe** [comment]
Seem like the issue come when receive 2 or more msg reset seq from acceptor. We will try to simulate this issue again.                                                                                                                                                                                                                                                                                                                                        -- Dung Vu | Developer | Amigo Quant Edge JSCmobile: +849 0896 6639email: dung.vu@quant-edge.comskype: dung.vtAmigo Quant Edge JSCSuite #503, IPH Building241 Xuan Thuy Street, Cau Giay DistHanoi, Vietnamtel: +84 (4) 7300 8999fax: +84 (4) 7300 3999web: www.quant-edge.comThis message (including attachments, if any) is confidential, may be privileged and is intended for the above-named recipient(s) only. If you have received this message in error, please notify me by return email and delete this message from your system. Any unauthorized use or disclosure of this message is strictly prohibited.Amigo Quant Edge JSC assumes no responsibility for errors, inaccuracies or omissions in these materials, and does not warrant the accuracy or completeness of the information contained within these materials. Amigo Quant Edge JSC shall not be liable for any special, indirect, incidental, or consequential damages, including without limitation losses, lost revenues, or lost profits that may result from these materials.                                                                                                                                                                                                                From: hungle-qeSent: Thursday, May 28, 2015 18:43To: connamara/quickfixnReply To: connamara/quickfixnCc: Dung VuSubject: Re: [quickfixn] Incorrect handling of SequenceRest Gap Fill Message (#309)message log client side:

20150527-09:45:37.052 : 8=FIX.4.4�9=91�35=0�34=4025�49=Client�52=20150527-09:45:37.052�56=Server�112=Appia-20150527-09:45:40�10=160�
20150527-09:45:40.313 : 8=FIX.4.4�9=235�35=D�34=4026�49=Client�52=20150527-09:45:40.313�56=Server�1=CS555A�11=197360340051�15=USD�21=1�22=5�38=4�40=2�44=126.3�48=1730�49=Client�54=2�55=KC�59=0�107=3�115=�116=Client�200=201507�207=IEPA�461=FXXXXX�513=TCB038001�10=163�
20150527-09:46:12.043 : 8=FIX.4.4�9=75�35=A�34=4027�49=Client�52=20150527-09:46:12.043�56=Server�98=0�108=10�10=111�
20150527-09:46:42.463 : 8=FIX.4.4�9=75�35=A�34=4028�49=Client�52=20150527-09:46:42.463�56=Server�98=0�108=10�10=121�
20150527-09:46:46.691 : 8=FIX.4.4�9=235�35=D�34=4029�49=Client�52=20150527-09:46:46.691�56=Server�1=CS555A�11=197360406445�15=USD�21=1�22=5�38=3�40=2�44=1633�48=1779�49=Client�54=1�55=LRC�59=0�107=3�115=�116=Client�200=201511�207=IEPA�461=FXXXXX�513=TCB011001�10=227�
20150527-09:47:12.883 : 8=FIX.4.4�9=75�35=A�34=4030�49=Client�52=20150527-09:47:12.883�56=Server�98=0�108=10�10=118�
20150527-09:47:43.303 : 8=FIX.4.4�9=75�35=A�34=4031�49=Client�52=20150527-09:47:43.303�56=Server�98=0�108=10�10=110�
20150527-09:48:13.723 : 8=FIX.4.4�9=75�35=A�34=4032�49=Client�52=20150527-09:48:13.723�56=Server�98=0�108=10�10=115�
20150527-09:48:13.801 : 8=FIX.4.4�9=0071�35=A�49=Server�56=Client�34=4119�52=20150527-09:48:17�98=0�108=10�10=015�
20150527-09:48:13.801 : 8=FIX.4.4�9=75�35=2�34=4033�49=Client�52=20150527-09:48:13.801�56=Server�7=4115�16=0�10=096�
20150527-09:48:13.801 : 8=FIX.4.4�9=0074�35=2�49=Server�56=Client�34=4120�52=20150527-09:48:17�7=4022�16=4031�10=142�
20150527-09:48:13.801 : 8=FIX.4.4�9=108�35=4�34=4022�43=Y�49=Client�52=20150527-09:48:13.801�56=Server�122=20150527-09:48:13.801�36=4026�123=Y�10=032�
20150527-09:48:13.801 : 8=FIX.4.4�9=266�35=D�34=4026�43=Y�49=Client�52=20150527-09:48:13.801�56=Server�115=�116=Client�122=20150527-09:45:40.313�1=CS555A�11=197360340051�15=USD�21=1�22=5�38=4�40=2�44=126.3�48=1730�54=2�55=KC�59=0�107=3�200=201507�207=IEPA�461=FXXXXX�513=TCB038001�10=175�
20150527-09:48:13.801 : 8=FIX.4.4�9=108�35=4�34=4027�43=Y�49=Client�52=20150527-09:48:13.801�56=Server�122=20150527-09:48:13.801�36=4029�123=Y�10=040�
20150527-09:48:13.801 : 8=FIX.4.4�9=266�35=D�34=4029�43=Y�49=Client�52=20150527-09:48:13.801�56=Server�115=�116=Client�122=20150527-09:46:46.691�1=CS555A�11=197360406445�15=USD�21=1�22=5�38=3�40=2�44=1633�48=1779�54=1�55=LRC�59=0�107=3�200=201511�207=IEPA�461=FXXXXX�513=TCB011001�10=239�
20150527-09:48:13.801 : 8=FIX.4.4�9=108�35=4�34=4030�43=Y�49=Client�52=20150527-09:48:13.801�56=Server�122=20150527-09:48:13.801�36=4032�123=Y�10=028�
20150527-09:48:13.833 : 8=FIX.4.4�9=0100�35=4�49=Server�56=Client�34=4115�43=Y�122=20150527-09:48:17�52=20150527-09:48:17�36=4121�123=Y�10=193�
20150527-09:48:13.833 : 8=FIX.4.4�9=0105�35=3�49=Server�56=Client�34=4121�52=20150527-09:48:17�45=4026�58=Framing error: Checksum not found.�10=117�
20150527-09:49:03.909 : 8=FIX.4.4�9=75�35=A�34=4035�49=Client�52=20150527-09:49:03.877�56=Server�98=0�108=10�10=128�
20150527-09:49:03.940 : 8=FIX.4.4�9=0071�35=A�49=Server�56=Client�34=4122�52=20150527-09:49:07�98=0�108=10�10=009�
20150527-09:49:03.971 : 8=FIX.4.4�9=0074�35=2�49=Server�56=Client�34=4123�52=20150527-09:49:07�7=4026�16=4034�10=152�
20150527-09:49:03.971 : 8=FIX.4.4�9=108�35=4�34=4026�43=Y�49=Client�52=20150527-09:49:03.971�56=Server�122=20150527-09:49:03.971�36=4035�123=Y�10=052�
20150527-09:49:04.080 : 8=FIX.4.4�9=235�35=D�34=4036�49=Client�52=20150527-09:49:04.080�56=Server�1=CS555A�11=197360539858�15=USD�21=1�22=5�38=1�40=2�44=150.5�48=1730�49=Client�54=2�55=KC�59=0�107=3�115=�116=Client�200=201507�207=IEPA�461=FXXXXX�513=TCB038001�10=190�

—Reply to this email directly or view it on GitHub.

**Comment by dungvu-qe** [comment]
After 2-3 request reset msg: the initator store.get() return list msg with count = 0. We dont know why, and try to re-make this case again.

**Comment by oclancy** [comment]
Original report is from a while ago but I have recreated this issue.

Basically as OP states the GapFill=Y causes the SequenceReset Verify call to fail:

Below the ResendRequest was for 17907 to 0 (last seen message was 19102), Next expected Sequence number is therefore 19103.

The last 4 messages of the resend are (trade info redacted):
`20230712-15:37:43.309 : 8=FIX.4.2 | 9=416 | 35=8 | 122=20230712-15:32:14.946 | 56=COMPID1 | 52=20230712-15:32:40.709 | 49=COMPID2 | 43=Y | 34=19101 | 120=USD | 151=xxxx | 150=1 | 39=1 | 38=xxxx | 37=xxxx | 32=xx | 167=CS | 31=xxxx | 
60=20230712-15:32:14.867 | 17=xxxx | 48=xxxx | 15=xxxx | 14=xx | 76=xxxx | 44=0 | 75=20230712 | 11=xxxxx | 40=1 | 7504=xxxx | 59=0 | 7500=xxxx | 207=x | 6=xxxxx | 55=xxxx | 54=2 | 22=5 | 1=xxxx | 21=2 | 20=0 | 10=242 | `

19103 was an admin message so gap fill sent to cover and set next seq num to 19104
`20230712-15:37:43.695 : 8=FIX.4.2 | 9=103 | 35=4 | 52=20230712-15:32:40.709 | 49=COMPID2 | 34=19102 | 56=COMPID1 | 43=Y | 36=19104 | 123=Y | 10=145 |` 

Next msg (19104) received
`20230712-15:37:43.852 : 8=FIX.4.2 | 9=417 | 35=8 | 122=20230712-15:32:39.040 | 56=COMPID1 | 52=20230712-15:32:40.709 | 49=COMPID2 | 43=Y | 34=19104 | 120=USD | 151=xxxx | 150=1 | 39=1 | 38=xxxx | 37=xxxx | 32=xx | 167=CS | 31=xxxx | 60=20230712-15:32:15.562 | 17=xxxxx | 48=xxxx | 15=xxxx | 14=xx | 76=xxxx | 44=0 | 75=20230712 | 11=xxxxx | 40=1 | 7504=xxxx | 59=0 | 7500=xxxx | 207=x | 6=xxxxx | 55=xxxx | 54=5 | 22=5 | 1=xxxx | 21=2 | 20=0 | 10=144 |` 

Due to fault in Session.Verify  for SequenceReset messages,  NewSeqNo will not be set (from SequenceReset message above), and a new ResendRequest is generated. 
`20230712-15:37:44.281 : 8=FIX.4.2 | 9=95 | 35=2 | 34=1376 | 49=COMPID1 | 52=20230712-15:37:44.096 | 56=COMPID2 | 7=**19103** | 16=0 | 10=188 | `

The next SequenceReset message received will actually have the next expected msgseqnum (19103 in this case) and Session.Verify will return true at which point then the NewSeqNo(36) will be acted upon.

The problem I have is that with very heavy message flow, once the first resend is satisfied, the Sender then continues to send stored up messages at which point it then recieves the second ResendRequest (generated in error) and causes a large cycle of resends which can kill the connection.

I will create a PR if I can rationalise a fix

**Comment by gbirchmeier** [comment]
Can someone tell me if these logs were on CME or not?

**Comment by oclancy** [comment]
Hi @gbirchmeier, no these logs were not from CME. I think the issue was found on Bloomberg and recreated locally.

**Comment by gbirchmeier** [comment]
Great to know, thanks for responding.

**Comment by oclancy** [comment]
No worries. I've also remembered it was Tora (Now LSE TORA) not Bloomberg. (Apologies, this was a while ago)

**Comment by gbirchmeier** [comment]
This whole section of the code is not fun to mess with.  Furthermore, the original report is 10 years old, so... argh.  Hard to follow up on that.  And the log someone included above was useless.

But the initial submitter's description, I see two separate quirks to answer:

1. The Gap Fill message's 36/NewSeqNo exceeds that of the request that it is responding to (which was Begin/End: 15260/17759).
    * I don't see this by itself as a problem for most counterparties.  The QF/n app can pick up at that new seq no and continue accordingly.  Per my reading of the FIX spec, this would be the correct behavior.
    * (I do know a counterparty that wants different behavior in this situation.  Support will be added in the future for this, but that is out of scope for this issue.)

2. "QUICK FIX ignores the next expected resend request itself" - Yes, that's totally a bug.

**Now my read on oclancy's comment:**

(COMPID2 is the side requesting resend and then messing up)

* Last seen message is 19102
* Send ResendRequest from 17907-0 (0 is infinity)

```
# 19101 from COMPID2
8=FIX.4.2|9=416|35=8|122=20230712-15:32:14.946|56=COMPID1|52=20230712-15:32:40.709|49=COMPID2|43=Y|34=19101...
# gap fill sets next to 19104
8=FIX.4.2|9=103|35=4|52=20230712-15:32:40.709|49=COMPID2|34=19102|56=COMPID1|43=Y|36=19104|123=Y|10=145|

# COMPID2 didn't properly update NextExpectedTargetSeqNum to 19104

# 19104 from COMPID2
8=FIX.4.2|9=417|35=8|122=20230712-15:32:39.040|56=COMPID1|52=20230712-15:32:40.709|49=COMPID2|43=Y|34=19104...

# COMPID2 thinks it missed 10103, but it didn't, and this is the bug
8=FIX.4.2|9=95|35=2|34=1376|49=COMPID1|52=20230712-15:37:44.096|56=COMPID2|7=**19103**|16=0|10=188|
```

**Comment by gbirchmeier** [comment]
@oclancy I'm having trouble following `TestSequenceResetWithGapFillAndPossDupFlagTrueDuringResendRequestIsProcessed`.

I understand the code in the test, but I don't understand what this scenario is:

1. It doesn't match the scenario in your comment
2. It doesn't look like it demonstrates a misbehavior -- the Logon/seq=10 *should* be "replayed" (setting nextExpectedSeq=11), and the SeqReset/seq=10 *should* be ignored or even rejected, because 10 is lower than the expected 11.  (Plus, why would the counterparty send a Logon/10 followed by a SeqReset/10?  That doesn't make sense.  Two different messages can't have the same seq num!)

Regarding your fix:  It breaks AcceptanceTest `10_MsgSeqNumLess.def`.  It's not impossible that the ATs might be wrong, but it's going to take a strong case to convince me of that.

I'm going to attempt a UT that more closely reflects the situation in your comment and see how that shakes out.

**Comment by gbirchmeier** [comment]
@oclancy I see it now!  In writing my own test to reflect your scenario, I'm finally spotting the problem (and my misconceptions).  Your test was accurate, though the second Logon threw my brain off course.

Corrections to my observations about your test:

* The seq num of the SequenceReset is supposed to follow the seq increment of the resend range.  The SeqReset is seq=10 because the test just resent seq=9.
* The original 10 is a Logon, which is an admin message, and all admin messages should be replaced with a gapfill (i.e. a SeqReset with 123/GapFillFlag=Y).

And *now* I see what makes this situation obscure-- it's the skip after 10.  It's reasonable to expect that the scenario would have an 11 message in there somewhere, but you are seeing a scenario where it gets skipped entirely.

1. Client receives an admin seq=10 that causes a resend from n to infinity
2. remote fulfills the resend until 9
3. (Client processes 10 off its queue right after it processes 9; it's next expected seq is now 11)
4. remote sends gapfill that starts at 10 (the admin message)... **and ends at 12**
5. **Client discards the gapfill as a dupe of the 10 that it already processed.  It still waits for 11.  If it gets a 12, it will erroneously ResendRequest for 11.**

I have written my own version of this test.  It can be pasted in to SessionTest:

```
    [Test]
    public void TestQueuedMessageAfterGapFill()
    {
        Assert.That(_session!.IgnorePossDupResendRequests, Is.EqualTo(false));
        //_session.RequiresOrigSendingTime = false; // default is true
        Logon();
        SendNOSMessage(); // seq 2

        QuickFix.FIX42.Heartbeat hb = new();
        SendTheMessage(hb, 5); // seq=5, causes ResendRequest to be sent

        // Verify the ResendRequest that was just sent
        var resendRequest =
            (QuickFix.FIX42.ResendRequest)_responder.MsgLookup[QuickFix.Fields.MsgType.RESEND_REQUEST].Dequeue();
        Assert.That(resendRequest.BeginSeqNo.Value, Is.EqualTo(3));
        Assert.That(resendRequest.EndSeqNo.Value, Is.EqualTo(0));

        Assert.That(_session.NextTargetMsgSeqNum, Is.EqualTo(3));

        // Resends
        SendNOSMessage(3, possDupFlag: true);
        SendNOSMessage(4, possDupFlag: true);
        // Now the session processes the Heartbeat/seq=5 from the queue.
        // Next expected = 6.

        // But the counterparty still needs to resend the seq=5!
        // The 5 is a Heartbeat, an admin message, so it's a gapfill.
        // The gapfill goes to 7, omitting ever sending a 6.
        QuickFix.FIX42.SequenceReset sr = new(new QuickFix.Fields.NewSeqNo(7));
        sr.Header.SetField(new QuickFix.Fields.PossDupFlag(true));
        sr.SetField(new QuickFix.Fields.GapFillFlag(true));
        SendTheMessage(sr, 5);

        // Even though client already processed the original Heartbeat/seq=5,
        //   it must not discard the GapFill 5.
        Assert.That(_session.NextTargetMsgSeqNum, Is.EqualTo(7));
    }
```

**Comment by gbirchmeier** [comment]
The problem with the fix in PR 784 is that it breaks test 10_MsgSeqNumLess.def.  In this test, the counterparty sends SeqResets that *should* be ignored.  But in that test, there is never a SeqReset that actually gets underway.

I think... maybe the best we can do here is:

1) When a SeqReset/GapFill is received that is too low
    a) if ResendRequest is underway (i.e. engine has sent a ResendReq that is not finished), obey it
    b) if not underway, ignore it (unlikely)

This might be sufficient.  I'm going to see if I can do it.

**Comment by gbirchmeier** [comment]
My fix is in #957.  My previous comment's approach wasn't quite right, as the ResendRequest is already "completed" when this glitch happens.

I had to introduce a state variable to note when RR is completed by a queue message.  When this is true, **and only when this is true**, then we can have a special behavior to obey the "dupe" SeqReset.

**Comment by jzabroski** [comment]
@gbirchmeier Just wondering if AI helped identify the the bug in this 10 year old bug?

**Comment by oclancy** [comment]
> [@gbirchmeier](https://github.com/gbirchmeier) Just wondering if AI helped identify the the bug in this 10 year old bug?

It did not

**Comment by gbirchmeier** [comment]
@jzabroski no
